refactor ceres nonlinear approach

Alpus opened the pull request:
* Add class ModelFitter
* Add multi-camera estimation support
* Simplify ceres example

patrikhuber committed:
Added eos::core::Image resize function

This has been added by @PhilippKopp, largely copied from the interpolation code in the texture extraction I think.

The code for Image3u and Image4u is largely duplicated, we should templatise the method and get rid of that code repetition.

patrikhuber committed:
Uncommented v2::extract_texture and updated all affected code

* Removed cv::Vec and cv::Mat across all affected code, i.e. the texturing and the "v2" SoftwareRenderer
* Reworked the Texture class to use eos::core::Image instead of cv::Mat
* Prefixed header guards with EOS_ (the ones that weren't yet)

With these changes, `v2::extract_texture` compiles again and it extracts some texture without crashing. I haven't really tested it more yet though.

patrikhuber committed:
Updated SoftwareRenderer to compile again with the rest of the code

There are some unnecessary conversions going on, where we create glm::vec objects, because the rendering code uses those, while the mesh's vertices and color values are Eigen::Vector types.
We should eventually move all these glm types to Eigen.

Updated this class from cv::Mat to eos::core::Image as well.

patrikhuber committed:
Change rasterizer member variable in SoftwareRenderer from unique_ptr to stack-allocated

I think I originally used a unique_ptr there because Rasterizer is not default-constructible, but of course we can still use a stack-allocated member, if we initialise it directly in the constructor.

patrikhuber committed:
Removed outdated todo comment

patrikhuber committed:
Updated TexturingFragmentShader to compile again

patrikhuber committed:
Updated documentation of the FragmentShaders to specify the colour is always in the range [0, 1]

patrikhuber committed:
Change uchar to unsigned char

'uchar' was defined in OpenCV headers and is of course not a standard C++ type.

patrikhuber committed:
Pass Image by const& in v2::extract_texture

patrikhuber committed:
Ordered includes

patrikhuber committed:
Ran clang-format on Texture.hpp

Alpus committed:
Refactor ceres approach

Alpus committed:
Add IntelliJ IDEA files in .gitignore

s1ddok commented on include/eos/morphablemodel/MorphableModel.hpp:
```diff
@@ -353,7 +356,8 @@ class MorphableModel
             mesh = sample_to_mesh(shape_sample, color_sample, shape_model.get_triangle_list(),
-                                  color_model.get_triangle_list(), texture_coordinates);
+                                  color_model.get_triangle_list(), texture_coordinates,
+                                  texture_triangle_indices);
```
it feels a bit inconsistent, that there is triangle list everywhere but here precomputed indices are used

Alpus replied:
This is commit from origin repo

s1ddok commented on include/eos/fitting/ceres_nonlinear.hpp:
```diff
@@ -30,29 +33,46 @@
+#include "ceres/problem.h"
 
+
+#ifndef EOS_CERES_USE_OPENCV
```
strange naming, why `CERES`?

Alpus replied:
It prevent opencv usage only for ceres approach

Alpus committed:
Apply clang formatting

Alpus committed:
Add PR suggestions

Alpus committed:
Remove opencv usage from ceres approach

Alpus committed:
Fix compile error of the GCC with IndexLandmark braces initializer.

GCC compiler raise an error when it try to compile braces initializer of IndexLandmark
struct. This commit add explicit definition of intializers for this struct.

Alpus committed:
Fix a few bugs

Alpus committed:
Improve different model support

Alpus committed:
Fix fit-model-multi example

Alpus committed:
Remove default ceres options from solve member function

Alpus committed:
Fix forward declaration of get_vertex_colour function
